File: pixiejsClient/src/app/ecs/components/ParentChildComponent.ts
import { Component, component, serverField } from "../core/Component";
import { NTT } from "../core/NTT";
import { ServerComponentType } from "../../enums/ComponentIds";

export interface ParentChildConfig {
  parentId: string;
  gridX?: number;
  gridY?: number;
  shape?: number;
  rotation?: number;
}

@component(ServerComponentType.ParentChild)
export class ParentChildComponent extends Component {
  // changedTick is inherited from Component base class
  @serverField(1, "guid") parentId: string;
  @serverField(2, "i8") gridX: number;
  @serverField(3, "i8") gridY: number;
  @serverField(4, "u8") shape: number;
  @serverField(5, "u8") rotation: number;

  constructor(ntt: NTT, config?: ParentChildConfig | string) {
    super(ntt);

    if (config) {
      if (typeof config === "string") {
        this.parentId = config;
        this.gridX = 0;
        this.gridY = 0;
        this.shape = 0;
        this.rotation = 0;
      } else {
        this.parentId = config.parentId;
        this.gridX = config.gridX || 0;
        this.gridY = config.gridY || 0;
        this.shape = config.shape || 0;
        this.rotation = config.rotation || 0;
      }
    } else {
      // Defaults for deserialization
      this.parentId = "00000000-0000-0000-0000-000000000000";
      this.gridX = 0;
      this.gridY = 0;
      this.shape = 0;
      this.rotation = 0;
    }
  }
}

File: server/Simulation/Components/ParentChildComponent.cs
using System.Runtime.InteropServices;
using server.ECS;
using server.Enums;

namespace server.Simulation.Components;

[Component(ComponentType = ComponentType.ParentChild, NetworkSync = true)]
[StructLayout(LayoutKind.Sequential, Pack = 1)]
public struct ParentChildComponent(NTT parentId, sbyte gridX = 0, sbyte gridY = 0, byte shape = 0, byte rotation = 0)
{
    /// <summary>
    /// Tick when this component was last changed, used for network sync.
    /// MUST be first field for raw byte access in ComponentSerializer.
    /// </summary>
    public long ChangedTick = NttWorld.Tick;

    /// <summary>
    /// Entity ID of the parent entity
    /// </summary>
    public NTT ParentId = parentId;

    /// <summary>Grid X position in ship builder grid</summary>
    public sbyte GridX = gridX;
    /// <summary>Grid Y position in ship builder grid</summary>
    public sbyte GridY = gridY;
    /// <summary>Shape type (1=triangle, 2=box, etc.)</summary>
    public byte Shape = shape;
    /// <summary>Rotation in 90° increments (0-3 = 0°, 90°, 180°, 270°)</summary>
    public byte Rotation = rotation;
}